service/src/main/java/io/camunda/service/AdHocSubprocessActivityServices.java

```diff
@@ -99,6 +112,50 @@ public List<AdHocSubprocessActivity> findActivatableActivities(
         "Failed to find Ad-Hoc Subprocess with ID '%s'".formatted(adHocSubprocessId));
   }
 
+  public CompletableFuture<ProcessInstanceModificationRecord> activateActivities(
+      final Long adHocSubprocessInstanceKey, final List<String> flowNodeIds) {
+    final var flowNode = flowNodeInstanceServices.getByKey(adHocSubprocessInstanceKey);
+
+    // TODO type returns as UNKNOWN - something to fix in exporter?
+    /*
+    if (flowNode.type() != FlowNodeType.AD_HOC_SUB_PROCESS) {
+      throw new IllegalArgumentException(
+          "Flow node instance with key '%s' is not an Ad-Hoc Subprocess"
+              .formatted(adHocSubprocessInstanceKey));
+    }
+    */
+
+    final var activatable =
+        findActivatableActivities(flowNode.processDefinitionKey(), flowNode.flowNodeId()).stream()
+            .map(AdHocSubprocessActivity::flowNodeId)
+            .toList();
+
+    final var unsupported =
+        flowNodeIds.stream().filter(flowNodeId -> !activatable.contains(flowNodeId)).toList();
+    if (!unsupported.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Flow node IDs '%s' are not activatable in Ad-Hoc Subprocess with ID '%s'"
+              .formatted(unsupported, flowNode.flowNodeId()));
+    }
+
+    final var activateInstructions =
+        flowNodeIds.stream()
+            .map(
+                flowNodeId -> {
+                  final var instruction = new ProcessInstanceModificationActivateInstruction();
+                  instruction.setAncestorScopeKey(-1);
+                  instruction.setElementId(flowNodeId);
+                  return instruction;
```

Comment:
This is the same logic the [Modify process instance API](https://docs.camunda.io/docs/apis-tools/camunda-api-rest/specifications/modify-process-instance/) does when calling it with:

```json5
// POST http://localhost:8080/v2/process-instances/2251799813685271/modification
{
    "activateInstructions": [
        {
            "elementId": "fooCalculation"
        },
        {
            "elementId": "barCalculation"
        }
    ]
}
```